### PR TITLE
Implement support for chunked writing mode for querylogs.sql…

### PR DESCRIPTION
**Pull request by cyulysses-corp**
…when target output location is a file directory.

### CODE REVIEW

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImpl.java [hunk 1]
```diff
@@ -73,12 +158,41 @@ public void executeScript(
...
+      ImmutableList<String> sortingColumns) {
     Preconditions.checkArgument(
         scriptsMap.containsKey(scriptName),
         String.format("Script name %s is not available.", scriptName));
-    return sqlTemplateRenderer.renderTemplate(scriptName, scriptsMap.get(scriptName).get());
+    String rawScript = scriptsMap.get(scriptName).get();
+    if (!sortingColumns.isEmpty()) {
+      String orderClause =
+          sortingColumns.stream()
+              .collect(Collectors.joining(", ", "\nORDER BY ", " ASC NULLS FIRST"));
```
Comment (haselbach) [inline src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImpl.java:192]: I think it would be cleaner to provide the order columns as a parameter to the template and have the template create the `ORDER BY` part. This goes against the idea of using templates in the first place.

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImpl.java [hunk 2]
```diff
@@ -73,12 +158,41 @@ public void executeScript(
...
+        instant
+            .truncatedTo(ChronoUnit.SECONDS)
+            .toString()
+            .replaceAll("[\\-:]", "")
+            .replaceAll("[Z]", "S")
+            .trim();
+    // Keep 6 digits of the fractional seconds from nano, which is equivalent to round(nanos / 10^9
+    // * 10^6) with left-padded zeros.
+    String sixDigitNanos =
+        String.format((Locale) null, "%06d", Math.round(timestamp.getNanos() / Math.pow(10, 3)));
```
[the remark this answers is not in the document]
Comment (cyulysses-corp) [reply src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImpl.java:176]: Per https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#format-java.lang.String-java.lang.Object...-, the locale would be Locale.getDefault(). With locale set to null, we avoid locale-related differences by not involving any localization (https://docs.oracle.com/javase/8/docs/api/java/util/Formatter.html#Formatter-java.util.Locale-)

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImpl.java [hunk 3]
```diff
@@ -50,17 +65,87 @@ public void executeScript(
...
+    oneChunk:
+    try (ResultSetRecorder<GenericRecord> dumper =
+        AvroResultSetRecorder.create(
+            schema, dataEntityManager.getEntityOutputStream(tempFileName))) {
+      for (int i = 0; i < chunkRows; i++) {
+        // Process first, then advance the row.
+        latestTimestamp = resultSet.getTimestamp(labelColumn);
+        dumper.add(parseRowToAvro(resultSet, schema));
+        if (!resultSet.next()) {
+          break oneChunk;
```
[the remark this answers is not in the document]
Comment (cyulysses-corp) [reply src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImpl.java]: You're right. Deleted it.

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImpl.java [hunk 4]
```diff
@@ -73,12 +158,41 @@ public void executeScript(
...
+    Instant instant = timestamp.toInstant();
+    // Replace the "Z" that signifies UTC timezone with "S" that signifies the separator between
+    // seconds and fractional seconds; remove separators that are not necessary for human
+    // interpretation.
+    String instantWithoutNanoseconds =
+        instant
+            .truncatedTo(ChronoUnit.SECONDS)
+            .toString()
+            .replaceAll("[\\-:]", "")
+            .replaceAll("[Z]", "S")
```
[the remark this answers is not in the document]
Comment (cyulysses-corp) [reply src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImpl.java]: The toString() method intrinsically uses DateTimeFormatter.ISO_INSTANT, which is both timezone-agnostic and locale-agnostic, thus I prefer using that one. I modified the string substitution so that only one regex scan is needed now.

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImpl.java (hunk 1 above)
Comment (cyulysses-corp) [reply src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImpl.java]: Good idea. Modified accordingly.

**Comment by cyulysses-corp** [comment]
(Re-examining whether the time formatting also has the problem that has been fixed in https://github.com/google/dwh-assessment-extraction-tool/pull/52/commits

File: src/javatests/com/google/cloud/bigquery/dwhassessment/extractiontool/dumper/DataEntityManagerTempTestImpl.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
+/*
```
Comment (frsv) [inline src/javatests/com/google/cloud/bigquery/dwhassessment/extractiontool/dumper/DataEntityManagerTempTestImpl.java:1]: This file seems really similar to DataEntityManagerTesting.java. How about creating a testing class, that can use either String or OutputStream under the hood?

Also, I believe it should be called Fake explicitly.
Consider moving it under `testing` directory to distinguish Fakes and Test suites

File: src/javatests/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImplTest.java [hunk 6]
```diff
@@ -123,27 +144,147 @@ public void getAllScriptNames_fail() throws Exception, SQLException {
...
+          String.format(
+              "INSERT INTO TestTable VALUES (%d, TIMESTAMP '2008-08-08 20:08:%02d.007000'"
+                  + " AT TIME ZONE INTERVAL '0:00' HOUR TO MINUTE)",
+              i, i + 8));
+    }
+    baseStmt.close();
+    connection.commit();
+    ImmutableList<String> expectedFiles =
+        ImmutableList.<String>builder()
+            .add("default_chunked-20080808T200808S007000-20080808T200810S007000_0.avro")
```
Comment (frsv) [inline src/javatests/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImplTest.java:226]: I'm curious, if there is a specific reason to use builder, instead of ImmutableList.ofI().

File: src/javatests/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImplTest.java [hunk 7]
```diff
@@ -123,27 +144,147 @@ public void getAllScriptNames_fail() throws Exception, SQLException {
...
+    DataFileReader<Record> reader;
+    reader =
+        getReaderFromAvroFile(
+            dataEntityManagerTmp.getAbsolutePath(
+                "default_chunked-20080808T200808S007000-20080808T200810S007000_0.avro"));
+    assertRecordEqualsExpected(reader.next(), 0, "2008-08-08T20:08:08.007000000Z");
+    assertRecordEqualsExpected(reader.next(), 1, "2008-08-08T20:08:09.007000000Z");
+    assertRecordEqualsExpected(reader.next(), 2, "2008-08-08T20:08:10.007000000Z");
+    assertFalse(reader.hasNext());
+    reader =
```
Comment (frsv) [inline src/javatests/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImplTest.java:260]: Mixing "act" and "assert" steps together is hard to follow.
Actually, to me it seems like two tests – one to check the first chunk and second to check the last

Also, splitting this test makes it very obvious, that "arrange" step should be moved to helper (part with creating and populating the table)

File: src/javatests/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImplTest.java [hunk 8]
```diff
@@ -44,37 +53,49 @@
...
+          () -> baseScript,
+          "default_chunked",
+          () ->
+              baseScript
+                  + "{{#if sortingColumns}}\n"
+                  + "ORDER BY {{#each sortingColumns}}{{this}}{{#unless"
+                  + " @last}},{{/unless}}{{/each}} ASC NULLS FIRST\n"
+                  + "{{/if}}");
+  private final ImmutableMap<String, ImmutableList<String>> sortingColumnsMap =
+      ImmutableMap.of("default_chunked", ImmutableList.of("TIMESTAMPS"));
```
Comment (frsv) [inline src/javatests/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImplTest.java:70]: from what I understand this actually is used only in `executeScript_writeChunked_success`. 
If it's true, it can be safely empty most of the time

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/subcommand/ExtractSubcommand.java [hunk 9]
```diff
@@ -199,6 +199,20 @@ private void setDryRun(boolean dryRun) {
     argumentsBuilder.setDryRun(dryRun);
   }
 
+  @Option(
+      names = "--rows-per-chunk",
+      defaultValue = "0",
+      description = {
+        "If larger than 0, the tool will attempt to use a chunked processing mode for scripts that"
+            + " support this, where the results for a supporting script are saved in chunks; this"
+            + " number defines the maximum rows per chunk holds. Currently, only querylogs.sql"
```
Comment (frsv) [inline src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/subcommand/ExtractSubcommand.java:208]: I'd rather exclude listing of scripts with feature support – it's really easy to forget to update this part, as it's not a part of the documentation

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImpl.java [hunk 10]
```diff
@@ -73,12 +158,35 @@ public void executeScript(
...
     Preconditions.checkArgument(
         scriptsMap.containsKey(scriptName),
         String.format("Script name %s is not available.", scriptName));
-    return sqlTemplateRenderer.renderTemplate(scriptName, scriptsMap.get(scriptName).get());
+    if (!sortingColumns.isEmpty()) {
+      sqlTemplateRenderer.getSqlScriptVariablesBuilder().setSortingColumns(sortingColumns);
+    }
+    String rendered =
+        sqlTemplateRenderer.renderTemplate(scriptName, scriptsMap.get(scriptName).get());
+    sqlTemplateRenderer.getSqlScriptVariablesBuilder().setSortingColumns(ImmutableList.of());
```
Comment (frsv) [inline src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImpl.java:188]: Did you consider sorting columns injection in the `getSqlTemplateRenderer` method?

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/subcommand/ExtractSubcommand.java (hunk 9 above)
Comment (cyulysses-corp) [reply src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/subcommand/ExtractSubcommand.java:208]: To clarify, do you suggest just removing the "Currently, only querylogs.sql..." part?

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/dbscripts/InternalScriptLoader.java [hunk 11]
```diff
@@ -21,38 +21,50 @@
...
-        .stream()
+    return scriptNames.stream()
         .collect(
             ImmutableMap.toImmutableMap(Functions.identity(), key -> scriptLoader(key + ".sql")));
   }
 
+  @Override
+  public ImmutableMap<String, ImmutableList<String>> getSortingColumnsMap() {
+    return new ImmutableMap.Builder<String, ImmutableList<String>>()
+        .put("querylogs.sql", ImmutableList.of("\"{{baseDatabase}}\".\"QryLogV\".\"StartTime\""))
```
[the remark this answers is not in the document]
Comment (cyulysses-corp) [reply src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/dbscripts/InternalScriptLoader.java:64]: vars.tableName wasn't there at first... will change it to make it compatible.

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/subcommand/ExtractSubcommand.java (hunk 9 above)
Comment (frsv) [reply src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/subcommand/ExtractSubcommand.java:208]: Yes

File: src/javatests/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImplTest.java (hunk 7 above)
Comment (cyulysses-corp) [reply src/javatests/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImplTest.java:260]: Per off-line discussion and post-discussion investigation, the conclusions are:

1. The "act" and "assert" steps are not mixed in this case; the naming of one helper function (getReaderFromAvroFile) however may cause such confusion, so I renamed this method.
2. The test is indeed too long with assertions on various aspects, thus I split it into two tests: one checking that all the files are present, the other checking that the contents in the first and last chunks are correct (with the assumption that the contents in the middle are unlikely to be incorrect in this case).

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImpl.java (hunk 10 above)
Comment (cyulysses-corp) [reply src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImpl.java:188]: I decided not to inject the sorting columns during getSqlTemplateRenderer based on these considerations:
1. sortingColumns serves the chunked-writing mechanism only and is hard-coded by design, i.e. should not be modifiable by the users, and thus should be separated from Vars of SqlScriptVariables.
2. For the same reason, sortingColumns should be tied to the ScriptLoader/ScriptManager instead of in ExtractExecutor.

It is true that resetting it to empty was not needed during executeScript because a new sqlTemplateRenderer is supplied every time; fixed that part.

File: src/javatests/com/google/cloud/bigquery/dwhassessment/extractiontool/dumper/DataEntityManagerTempTestImpl.java (hunk 5 above)
Comment (cyulysses-corp) [reply src/javatests/com/google/cloud/bigquery/dwhassessment/extractiontool/dumper/DataEntityManagerTempTestImpl.java:1]: Merged and renamed. Will move it, and other testing helper/utils to a `testing` directory in another CL.

File: src/javatests/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImplTest.java (hunk 6 above)
Comment (cyulysses-corp) [reply src/javatests/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImplTest.java:212]: Builder pattern felt more expandable at the time of writing...

File: src/javatests/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImplTest.java (hunk 8 above)
Comment (cyulysses-corp) [reply src/javatests/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImplTest.java:70]: This map being non-empty doesn't cause much overhead but can also verify that it does not interfere with other non-chunking scripts' executions.

Comment (frsv) [reply src/javatests/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImplTest.java:70]: I don't follow how does it interfere, It is a no-op for all non-chunking executions:
```
ImmutableList<String> sortingColumns = chunkMode ? sortingColumnsMap.get(scriptName) : ImmutableList.of();
```

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/dbscripts/InternalScriptLoader.java [hunk 12]
```diff
@@ -21,38 +21,50 @@
...
-        .stream()
+    return scriptNames.stream()
         .collect(
             ImmutableMap.toImmutableMap(Functions.identity(), key -> scriptLoader(key + ".sql")));
   }
 
+  @Override
+  public ImmutableMap<String, ImmutableList<String>> getSortingColumnsMap() {
+    return new ImmutableMap.Builder<String, ImmutableList<String>>()
+        .put("querylogs.sql", ImmutableList.of("\"StartTime\""))
```
Comment (frsv) [inline src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/dbscripts/InternalScriptLoader.java:64]: It depends on how will we approach sql logs and query texts merge, but just a note for the next PRs – we may want to sort sql_logs too and order of querylogs and sql_logs should be consistent to merge data effectively:
* StartTime/CollectTimeStamp ASC
* QueryID ASC
* (sql_logs specific) SqlRowNo DESC

**Review by frsv — APPROVED** [review]

File: src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/dbscripts/InternalScriptLoader.java (hunk 12 above)
Comment (cyulysses-corp) [reply src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/dbscripts/InternalScriptLoader.java:64]: Thanks for the heads-up - I'll keep that in mind.

File: src/javatests/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImplTest.java (hunk 8 above)
Comment (cyulysses-corp) [reply src/javatests/com/google/cloud/bigquery/dwhassessment/extractiontool/db/ScriptManagerImplTest.java:70]: Indeed it is, with the current implementation details. Testing is also for when such details are changed.
